ssdeep.go

```diff
@@ -137,66 +112,142 @@ type fuzzyReader interface {
 				// Finalize the hash string with the remaining data
-				sdeep.hashString1 += string(b64[sdeep.blockHash1%64])
-				sdeep.hashString2 += string(b64[sdeep.blockHash2%64])
+				state.hashString1 += string(b64[state.blockHash1%64])
+				state.hashString2 += string(b64[state.blockHash2%64])
 			}
 			break
 		}
 	}
-	return &FuzzyHash{
-		blockSize:    sdeep.blockSize,
-		hashString1:  sdeep.hashString1,
-		hashString2:  sdeep.hashString2,
-		fileLocation: fileLocation,
-	}, nil
+	return fmt.Sprintf("%d:%s:%s", state.blockSize, state.hashString1, state.hashString2), nil
 }
 
-// FuzzyFile hash of a provided reader
-func (sdeep *SSDEEP) FuzzyFile(f *os.File, fileLocation string) (*FuzzyHash, error) {
-	n, err := GetFileSize(f)
+// FuzzyFilename computes the fuzzy hash of a file.
+// FuzzyFilename will opens, reads, and hashes the contents of the file 'filename'.
+// It is the caller's responsibility to append the filename to the result after computation.
+// Returns an error when the file doesn't exist or ssdeep could not be computed on the file.
+func FuzzyFilename(filename string) (string, error) {
+	f, err := os.Open(filename)
 	if err != nil {
-		return nil, err
+		return "", err
 	}
+	defer f.Close()
 
-	return sdeep.fuzzyReader(f, n, fileLocation)
+	return FuzzyFile(f)
 }
 
-// FuzzyByte hash of a provided byte array
-func (sdeep *SSDEEP) FuzzyByte(blob []byte) (*FuzzyHash, error) {
-	n := len(blob)
-	br := bytes.NewReader(blob)
+// FuzzyFile computes the fuzzy hash of a file using os.File pointer.
+// FuzzyFile will computes the fuzzy hash of the contents of the open file, starting at the beginning of the file.
+// When finished, the file pointer is returned to its original position.
+// If an error occurs, the file pointer's value is undefined.
+// It is the callers's responsibility to append the filename to the result after computation.
+// Returns an error when ssdeep could not be computed on the file.
+func FuzzyFile(f *os.File) (string, error) {
+	currentPosition, err := f.Seek(0, io.SeekCurrent)
+	if err != nil {
+		return "", err
+	}
+
+	f.Seek(0, io.SeekStart)
+	state := newSsdeepState()
+	stat, err := f.Stat()
+	if err != nil {
+		return "", err
+	}
+	n := int(stat.Size())
+
+	result, err := state.fuzzyReader(f, n)
+	if err != nil {
+		return "", err
+	}
+
+	f.Seek(currentPosition, io.SeekStart)
+	return result, nil
+}
+
+// FuzzyBytes computes the fuzzy hash of a slice of byte.
+// It is the caller's responsibility to append the filename, if any, to result after computation.
+// Returns an error when ssdeep could not be computed on the buffer.
+func FuzzyBytes(buffer []byte) (string, error) {
+	state := newSsdeepState()
+	n := len(buffer)
+	br := bytes.NewReader(buffer)
+
+	result, err := state.fuzzyReader(br, n)
+	if err != nil {
+		return "", err
+	}
+
+	return result, nil
+}
+
+
+// Hash interface related struct and functions
+type digest struct {
+	buffer []byte
+}
+
+func (d *digest) Write(p []byte) (n int, err error) {
```

Comment:
My concerns https://github.com/glaslos/ssdeep/issues/11#issuecomment-353777252 are regarding that the `Write()` interface is usually used to update the digest and discard the written data. In your implementation you are collecting data in memory on `Write` and then calculate the hash on `Sum`. As said, not really something we can fix as long as we don't know the final amount of data to be written.

Reply:
You are right, but ssdeep is not "classic" hash, so there is no much we can do. I'll be more clear in the documentation.

Reply:
Sounds good and I think the added information should be good for now. We can also add a note in the README later.

Reply:
It's a good idea. Can you update the README after the release with this caveat?